### PR TITLE
[ADD] IPPay Credit Card payment acquirer

### PR DESCRIPTION
**Pull request by dreispt**

### CODE REVIEW

File: payment_ippay/models/payment.py [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+            else:
+                return token
+
+    @api.model
+    def ippay_create(self, values, token_code=False):
+        # In case we already know the token assigned, just use it
+        token_code = token_code or self._ippay_get_token(values)
+        existing = self.sudo().search(
+            [
+                ("acquirer_ref", "=", token_code)(
```
Comment (mgosai) [inline payment_ippay/models/payment.py]: @dreispt :
Looks like missing ',' after tuple ("acquirer_ref", "=", token_code)
Should be ("acquirer_ref", "=", token_code), ("partner_id"*******)

File: payment_ippay/models/payment_transaction.py [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+                <TransactionType>SALE</TransactionType>
+                <TerminalID>%s</TerminalID>
+                <Token>%s</Token>
+                <TotalAmount>%s</TotalAmount>
+            </ippay>""" % (
+            acquirer.ippay_terminal_id,
+            self.payment_token_id.acquirer_ref,
+            str(self.amount).replace(".", ""),
+        )
+        # TODO_ add verbosity parameter?
```
Comment (mgosai) [inline payment_ippay/models/payment_transaction.py:31]: should this comment go?

File: payment_ippay/views/payment_view.xml [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+    <record id='acquirer_form_ippay' model='ir.ui.view'>
+        <field name='name'>acquirer.form.ippay></field>
+        <field name='model'>payment.acquirer</field>
+        <field name='inherit_id' ref='payment.acquirer_form'/>
+        <field name='arch' type='xml'>
+            <xpath expr="//group[@name='acquirer']" position="inside">
+                <group>
+                    <field name='api_url' widget="url" attrs="{'invisible': [('provider','!=', 'ippay')],
+                    'required': [('provider','=', 'ippay')]}"/>
```
Comment (mgosai) [inline payment_ippay/views/payment_view.xml]: required attribute already taken care in python field declaration, is this repetition ?

File: payment_ippay/views/payment_view.xml [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+                    <field name='ippay_terminal_id' attrs="{'invisible': [('provider','!=', 'ippay')],
+                    'required': [('provider','=', 'ippay')]}"/>
```
Comment (mgosai) [inline payment_ippay/views/payment_view.xml]: same here

File: payment_ippay/__manifest__.py [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+{
+    "name": "IPPay Payment Acquirer for Credit Cards",
+    "version": "12.0.1.0.0",
+    "license": "AGPL-3",
+    "summary": "Store and use credit cards using IPPay",
+    "author": "Open Source Integrators, Odoo Community Association (OCA)",
+    "maintainer": "Open Source Integrators",
+    "website": "http://www.opensourceintegrators.com",
+    "category": "Accounting",
+    "depends": ["payment", "website", "account_payment"],
```
Comment (mgosai) [inline payment_ippay/__manifest__.py]: account_payment already have payment in depends

File: payment_ippay/models/payment.py [hunk 6]
```diff
@@ -9,15 +10,22 @@
...
+        required_if_provider="ippay")
+    ippay_terminal_id = fields.Char(
+        "IPpay TerminalID",
+        required_if_provider="ippay")
 
     @api.model
     def ippay_s2s_form_process(self, data):
-        Token = self.env["payment.token"]
+        """Get Payment token ref from Ippay payment token."""
+        token = self.env["payment.token"]
```
Comment (dreispt) [inline payment_ippay/models/payment.py]: @hpatelserpentcs The convention is for object references to use Title uppercase names, like "Token". Lowercase "token" should be used for object instances (recordsets).

File: payment_ippay/models/payment.py [hunk 7]
```diff
@@ -30,29 +38,30 @@ def ippay_s2s_form_process(self, data):
...
+                payment_method = token.sudo().create(values)
         else:
-            PaymentMethod = Token.sudo().search(
+            payment_method = token.sudo().search(
                 [
                     ("id", "=", int(token_id)),
-                    ("partner_id", "=", values.get("partner_id")),
-                    ("acquirer_id", "=", values.get("acquirer_id")),
+                    ("partner_id", "=", data.get("partner_id")),
+                    ("acquirer_id", "=", data.get("acquirer_id")),
```
Comment (dreispt) [inline payment_ippay/models/payment.py]: Looks like we had a bug here.

File: payment_ippay/models/payment.py (hunk 6 above)
Comment (hpatelserpentcs) [reply payment_ippay/models/payment.py]: But it throws the flack8: variable in function should be lowercase

File: payment_ippay/models/payment.py (hunk 7 above)
Comment (hpatelserpentcs) [reply payment_ippay/models/payment.py]: values are inside 'if' condition and search call in else part

**Comment by dreispt** [comment]
@nikitavaghela We shouldn't need the `website` dependency. The `portal`module does not require it.
Do you think we can remove it?

**Comment by dreispt** [comment]
Pushed minor changes.

**Comment by hpatelserpentcs** [comment]
> @nikitavaghela We shouldn't need the `website` dependency. The `portal`module does not require it.
> Do you think we can remove it?

yes, you can remove the dependency "payment" and "website"

**Comment by dreispt** [comment]
@hpatelserpentcs Done, thanks!

**Comment by dreispt** [comment]
Superseded by #319

**Pull request closed without merging**